apps/passport-server/src/services/generic-issuance/pipelines/PretixPipeline.ts

```diff
@@ -1910,6 +1938,372 @@ export class PretixPipeline implements BasePipeline {
 
     return ids;
   }
+
+  /**
+   * Returns {@link MemberCriteria} for each of the ticket types this user has
+   * permission to check in.
+   */
+  private async ticketsUserCanCheckIn(
+    checkerEmail: string
+  ): Promise<MemberCriteria[]> {
+    const checkerTickets = [
+      ...(await this.db.loadByEmail(this.id, checkerEmail)),
+      ...(await this.getManualTicketsForEmail(checkerEmail))
+    ];
+
+    const permittedTickets: MemberCriteria[] = [];
+
+    for (const permission of this.definition.options.userPermissions ?? []) {
+      if (
+        checkerTickets.find((ticket) =>
+          ticketMatchesCriteria(ticket, permission.members)
+        )
+      ) {
+        permittedTickets.push(permission.canCheckIn);
+      }
+    }
+
+    // Return any event/product combinations for which the user has a superuser
+    // ticket.
+    return this.definition.options.events.reduce((memo, eventConfig) => {
```

Comment:
`flatMap` may be simpler to reason about here than `reduce`

Reply:
I've rewritten this to use `flatMap` and to use the undefined `productId` rather than listing out all of the supported products.